File: csharp/SegmentationEngine/OcrUtils.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace SegmentationEngine
{
    /// <summary>
    /// Pure OCR text utilities. 1:1 port of python/segmenter/ocr_utils.py.
    /// Only System.Text.RegularExpressions — no external dependencies.
    /// </summary>
    public static class OcrUtils
    {
        private static readonly Regex NonAlnum = new Regex("[^a-z0-9]+", RegexOptions.Compiled);
        private static readonly Regex Ws = new Regex("\\s+", RegexOptions.Compiled);

        public static string Normalize(string text)
        {
            if (text == null) return " ";
            string low = text.ToLowerInvariant();
            low = NonAlnum.Replace(low, " ");
            low = Ws.Replace(low, " ").Trim();
            return " " + low + " ";
        }

        public static string HeaderText(string text, double ratio, int minLines)
        {
            if (string.IsNullOrEmpty(text)) return text ?? "";
            string[] lines = text.Replace("\r\n", "\n").Split('\n');
            if (lines.Length == 0) return text;
            int k = Math.Max(minLines, (int)Math.Ceiling(lines.Length * ratio));
            if (k > lines.Length) k = lines.Length;
            var sb = new StringBuilder();
            for (int i = 0; i < k; i++)
            {
                if (i > 0) sb.Append('\n');
                sb.Append(lines[i]);
            }
            return sb.ToString();
        }

        public static bool ContainsPhrase(string normalizedHaystack, string normalizedPhrase)
        {
            return normalizedHaystack.IndexOf(normalizedPhrase, StringComparison.Ordinal) >= 0;
        }

        public static List<string> Tokenize(string normalizedText)
        {
            var outList = new List<string>();
            foreach (var t in normalizedText.Split(' '))
                if (t.Length > 0) outList.Add(t);
            return outList;
        }

        public static double Jaccard(List<string> a, List<string> b)
        {
            if ((a == null || a.Count == 0) && (b == null || b.Count == 0)) return 1.0;
            if (a == null || b == null || a.Count == 0 || b.Count == 0) return 0.0;
            var sa = new HashSet<string>(a);
            var sb = new HashSet<string>(b);
            int inter = 0;
            foreach (var x in sa) if (sb.Contains(x)) inter++;
            int union = sa.Count + sb.Count - inter;
            return union > 0 ? (double)inter / union : 0.0;
        }

        // ---- page number parsing (fuzzy, OCR-noise tolerant) ----
        private static readonly Regex[] PnPatterns = new[]
        {
            new Regex(@"page\s+(\d{1,3})\s+of\s+(\d{1,3})", RegexOptions.Compiled),
            new Regex(@"pg\.?\s*(\d{1,3})\s*/\s*(\d{1,3})", RegexOptions.Compiled),
            new Regex(@"\bpage\s*[:#]?\s*(\d{1,3})\s*/\s*(\d{1,3})", RegexOptions.Compiled),
            new Regex(@"\b(\d{1,3})\s*/\s*(\d{1,3})\b", RegexOptions.Compiled),
            new Regex(@"\bpage\s*[:#]?\s*(\d{1,3})\b", RegexOptions.Compiled),
        };

        public static void ParsePageNumber(string text, out int? current, out int? total)
        {
            current = null; total = null;
            if (string.IsNullOrEmpty(text)) return;
            string low = text.ToLowerInvariant();
            var candidates = new List<string> { low };
            if (low.Contains("page") || low.Contains("/"))
                candidates.Add(DenoiseDigits(low));

            foreach (var cand in candidates)
            {
                foreach (var pat in PnPatterns)
                {
                    var m = pat.Match(cand);
                    if (m.Success)
                    {
                        int cur;
                        if (!int.TryParse(m.Groups[1].Value, out cur)) continue;
                        int? tot = null;
                        if (m.Groups.Count >= 3 && m.Groups[2].Success)
                        {
                            int t;
                            if (int.TryParse(m.Groups[2].Value, out t)) tot = t;
                        }
                        if (cur <= 0 || cur >= 500) continue;
                        if (tot.HasValue && (tot.Value <= 0 || tot.Value >= 500)) continue;
                        if (tot.HasValue && cur > tot.Value) continue;
                        current = cur; total = tot; return;
                    }
                }
            }
        }

        private static string DenoiseDigits(string s)
        {
            // only obvious confusions: l/i/I->1, o/O->0, s/S->5, B->8, |->1
            var sb = new StringBuilder(s.Length);
            foreach (char c in s)
            {
                switch (c)
                {
                    case 'l': case 'i': case 'I': case '|': sb.Append('1'); break;
                    case 'o': case 'O': sb.Append('0'); break;
                    case 's': case 'S': sb.Append('5'); break;
                    case 'B': sb.Append('8'); break;
                    default: sb.Append(c); break;
                }
            }
            return sb.ToString();
        }

        // ---- business identifier extraction ----
        public static List<string> ExtractIds(string text, List<Regex> typeIdRegexes, List<Regex> genericRegexes)
        {
            var found = new List<string>();
            string low = (text ?? "").ToLowerInvariant();
            foreach (var pat in typeIdRegexes)
                foreach (Match m in pat.Matches(low))
                    found.Add(CanonId(m.Value));
            if (found.Count == 0)
                foreach (var pat in genericRegexes)
                    foreach (Match m in pat.Matches(low))
                        found.Add(CanonId(m.Value));

            var seen = new HashSet<string>();
            var outList = new List<string>();
            foreach (var f in found)
                if (f.Length > 0 && seen.Add(f)) outList.Add(f);
            return outList;
        }

        private static string CanonId(string match)
        {
            var sb = new StringBuilder();
            foreach (char c in match.ToLowerInvariant())
                if (char.IsLetterOrDigit(c)) sb.Append(char.ToUpperInvariant(c));
            return sb.ToString();
        }

        public static List<Regex> CompileList(IEnumerable<string> patterns)
        {
            var outList = new List<Regex>();
            foreach (var p in patterns) outList.Add(new Regex(p, RegexOptions.Compiled));
            return outList;
        }
    }
}

File: python/segmenter/ocr_utils.py
"""Pure-stdlib OCR text utilities: normalisation, header zone, page-number and
business-identifier parsing, and similarity. Everything here ports cleanly to
C# (System.Text.RegularExpressions + simple string ops).
"""
from __future__ import annotations

import math
import re
from typing import Dict, List, Optional, Tuple

_NON_ALNUM = re.compile(r"[^a-z0-9]+")
_WS = re.compile(r"\s+")

# Common OCR digit confusions, applied only to numeric-looking contexts.
_OCR_DIGIT_MAP = str.maketrans({"l": "1", "i": "1", "I": "1", "o": "0",
                                "O": "0", "s": "5", "S": "5", "B": "8", "|": "1"})


def normalize(text: str) -> str:
    """Lowercase, strip punctuation to spaces, collapse whitespace.
    Padded with spaces so phrase matching can use ' phrase ' word boundaries."""
    low = text.lower()
    low = _NON_ALNUM.sub(" ", low)
    low = _WS.sub(" ", low).strip()
    return " " + low + " "


def header_text(text: str, ratio: float, min_lines: int,
                header_words: int = 55) -> str:
    """Return the 'header zone' (the start of the page) — LINE-INDEPENDENT.

    TTA may deliver a page as normal multi-line text OR as one long flattened line
    with no newlines. To make both behave identically, the header zone is defined
    as the first `header_words` whitespace-delimited tokens of the page (not the
    first N lines). This way single-line and multi-line OCR of the same page yield
    the same header and therefore the same classification. (ratio / min_lines are
    accepted for backward compatibility but no longer used.)
    """
    if not text:
        return text
    parts = text.split()
    if len(parts) <= header_words:
        return text
    return " ".join(parts[:header_words])


def contains_phrase(normalized_haystack: str, normalized_phrase: str) -> bool:
    """Word-boundary-safe substring test on already-normalized strings."""
    return normalized_phrase in normalized_haystack


def tokenize(normalized_text: str) -> List[str]:
    return [t for t in normalized_text.split() if t]


def jaccard(a_tokens: List[str], b_tokens: List[str]) -> float:
    if not a_tokens and not b_tokens:
        return 1.0
    sa, sb = set(a_tokens), set(b_tokens)
    if not sa or not sb:
        return 0.0
    inter = len(sa & sb)
    union = len(sa | sb)
    return inter / union if union else 0.0


# --------------------------------------------------------------------------
# Page number parsing (fuzzy, OCR-noise tolerant). Returns (current, total).
# --------------------------------------------------------------------------
_PN_PATTERNS = [
    re.compile(r"page\s+(\d{1,3})\s+of\s+(\d{1,3})"),
    re.compile(r"pg\.?\s*(\d{1,3})\s*/\s*(\d{1,3})"),
    re.compile(r"\bpage\s*[:#]?\s*(\d{1,3})\s*/\s*(\d{1,3})"),
    re.compile(r"\b(\d{1,3})\s*/\s*(\d{1,3})\b"),
    re.compile(r"\bpage\s*[:#]?\s*(\d{1,3})\b"),
]


def parse_page_number(text: str) -> Tuple[Optional[int], Optional[int]]:
    low = text.lower()
    # light denoise of obvious "page l of S" style only around the word 'page'
    candidates = [low]
    if "page" in low or "/" in low:
        candidates.append(low.translate(_OCR_DIGIT_MAP))
    for cand in candidates:
        for i, pat in enumerate(_PN_PATTERNS):
            m = pat.search(cand)
            if m:
                cur = _safe_int(m.group(1))
                tot = _safe_int(m.group(2)) if m.lastindex and m.lastindex >= 2 else None
                if cur is not None and 0 < cur < 500 and (tot is None or 0 < tot < 500):
                    if tot is not None and cur > tot:
                        continue
                    return cur, tot
    return None, None


def _safe_int(s: str) -> Optional[int]:
    try:
        return int(s)
    except (TypeError, ValueError):
        return None


# --------------------------------------------------------------------------
# Business identifier extraction.
# --------------------------------------------------------------------------
def extract_ids(text: str, type_id_regexes: List[re.Pattern],
                generic_regexes: List[re.Pattern]) -> List[str]:
    low = text.lower()
    found: List[str] = []
    for pat in type_id_regexes:
        for m in pat.findall(low):
            found.append(_canon_id(m))
    if not found:  # only fall back to generic IDs if no typed id matched
        for pat in generic_regexes:
            for m in pat.findall(low):
                found.append(_canon_id(m))
    # de-dup, preserve order
    seen, out = set(), []
    for f in found:
        if f and f not in seen:
            seen.add(f)
            out.append(f)
    return out


def _canon_id(match) -> str:
    if isinstance(match, tuple):
        match = "".join(match)
    return re.sub(r"[^a-z0-9]", "", str(match)).upper()


def compile_list(patterns: List[str]) -> List[re.Pattern]:
    return [re.compile(p) for p in patterns]
